iroh-metrics/src/lib.rs

```diff
@@ -141,3 +153,97 @@ pub fn get_current_trace_id() -> TraceId {
         .span_context()
         .trace_id()
 }
+
+#[derive(PartialEq)]
+pub enum Collector {
+    #[cfg(feature = "gateway")]
+    Gateway,
+    #[cfg(feature = "resolver")]
+    Resolver,
+    #[cfg(feature = "bitswap")]
+    Bitswap,
+    #[cfg(feature = "store")]
+    Store,
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub async fn record<M>(c: Collector, m: M, v: u64)
+where
+    M: MetricType + std::fmt::Display,
+{
+    let r = CORE.read().await;
+    if r.enabled() {
+        match c {
+            #[cfg(feature = "gateway")]
+            Collector::Gateway => r.gateway_metrics().record(m, v),
+            #[cfg(feature = "resolver")]
+            Collector::Resolver => r.resolver_metrics().record(m, v),
+            #[cfg(feature = "bitswap")]
+            Collector::Bitswap => r.bitswap_metrics().record(m, v),
+            #[cfg(feature = "store")]
+            Collector::Store => r.store_metrics().record(m, v),
+            _ => panic!("not enabled/implemented"),
+        };
+    }
+}
+
+#[allow(unused_variables, unreachable_patterns)]
+pub fn record_sync<M>(c: Collector, m: M, v: u64)
```

Comment:
Some of the code using metrics is not async so I made this as something to discuss, but if it were to live on, I'd rather have the task spawning in the call rather than having the user have to spawn tasks every time to shoot off a metric. Easier to refactor later if we try other runtimes.